doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+
+```py
+>>> class versin(Function):
+...     def _eval_rewrite(self, rule, args, **hints):
+...         if rule == cos:
+...             return 1 - cos(*args)
+>>> versin(x).rewrite(cos)
+1 - cos(x)
+```
+
+Once we've defined this, {func}`~.simplify` is now able to simplify some
+expressions containing `versin`:
+
+```
+>>> from sympy import simplify
+>>> simplify(versin(x) + cos(x))
+1
+```
+
+<!--
+
+TODO: simplify() can be customized with _eval_simplify(), but it isn't very
+powerful right now (see https://github.com/sympy/sympy/issues/19281). Most
+users are better off just defining rewrite(), doit(), and/or expand().
+
+#### Simplify
+
+-->
+
+(custom-functions-doit)=
+#### `doit()`
+
+The {meth}`doit() <sympy.core.basic.Basic.doit>` method is used to evaluate
+"unevaluated" functions. To define `doit()` implement `doit(self, deep=True,
+**hints)`. If `deep=True`, `doit()` should recursively call `doit()` on the
+arguments. `**hints` will be any other keyword arguments passed to the user,
+which should be passed to any recursive calls to `doit()`. You can use `hints`
+to allow the user to specify specific behavior for `doit()`.
+
+The typical usage of `doit()` in custom `Function` subclasses is to perform more
+advanced evaluation which is not performed in [`eval()`](custom-functions-eval).
+
+For example, for our [`divides` example](custom-functions-divides-definition),
+there are several instances that could be simplified using some identities.
+For example, we defined `eval()` to evaluate on explicit integers, but we might
+also want to evaluate examples like `divides(k, k*n)` where the divisibility
+is symbolically true. One of the [best practices for
+`eval()`](custom-functions-eval-best-practices) is to avoid too much automatic
+evaluation. Automatically evaluating in this case might be considered too
+much, as it would make use of the assumptions system, which could be
+expensive. Furthermore, we might want to be able to represent `divides(k,
+k*n)` without it always evaluating.
+
+The solution is to implement these more advanced evaluations in `doit()`. That
+way, we can explicitly perform them by calling `expr.doit()`, but they won't
+happen by default. An example `doit()` for `divides` that performs this
+simplification (along with the [above definition of
+`eval()`](custom-functions-divides-eval)) might look like this:
+
+```
+>>> class divides(Function):
+...     # Define evaluation on basic inputs, as well as type checking that the
+...     # inputs are not nonintegral.
+...     @classmethod
+...     def eval(cls, m, n):
+...         # Evaluate for explicit integer m and n.
+...         if isinstance(m, Integer) and isinstance(n, Integer):
+...             return int(n % m == 0)
+...
+...         # For symbolic arguments, require m and n to be integer.
+...         if m.is_integer is False or n.is_integer is False:
+...             raise TypeError("m and n should be integers")
```

Comment:
Ideally we wouldn't use `eval` for type checking so that we can still check types when `evaluate=False`.